In [1]:
import numpy as np
import pandas as pd
import pickle

from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV
from sklearn.preprocessing import MinMaxScaler
from sklearn.utils import resample
from sklearn.metrics import classification_report,accuracy_score

from sklearn.svm import SVC
from sklearn.multioutput import MultiOutputClassifier

from sklearn.linear_model import SGDClassifier
from sklearn.neighbors import KNeighborsClassifier

In [2]:
    with open('features.pkl', 'rb') as file:
        X = pickle.load(file)
        X=np.array(X)

    with open('labels.pkl', 'rb') as file:
        Y = pickle.load(file)
        Y=np.array(Y)

In [31]:
columns_y=['y0','y1','y2','y3','y4','y5','y6','y7','y8','y9']
y_df=pd.DataFrame(Y,columns=columns_y)
    
columns_x=['age','gender','income','became_member_on']
x_df=pd.DataFrame(X,columns=columns_x)
    
for col in columns_y:
    
    print(col)
    print('\n')
    table=pd.concat([x_df,y_df[col]],axis=1)
    df_majority = table[table[col]==0]
    df_minority = table[table[col]==1]
    
    # Upsample minority class
    df_minority_upsampled = resample(df_minority, 
                                     replace=True,   # sample with replacement
                                     n_samples=df_majority.shape[0]  # to match majority class
                                ) 
 
    # Combine majority class with upsampled minority class
    table = pd.concat([df_majority, df_minority_upsampled])
    
    scaler=MinMaxScaler()
    rescaled_data=scaler.fit_transform(table[columns_x])
    X_train, X_test, y_train, y_test = train_test_split(rescaled_data,\
                                                        table[col], test_size=0.2)

    y_test=np.array(y_test)
    y_train=np.array(y_train)
    
    #table[col].value_counts()
    
    clf = SVC(kernel='rbf',class_weight='balanced',C=1, gamma='auto')

    #clf = KNeighborsClassifier(n_neighbors=1)

    clf.fit(X_train, y_train)
    
    #compute y_predicted
    y_pred = clf.predict(X_test)
    #compute accuracy
    accuracy = (y_pred == y_test).mean()
    print('Accuracy: {}'.format(accuracy))
    print('\n')
    print(classification_report(y_test,y_pred))

y0


Accuracy: 0.5803820135296458


              precision    recall  f1-score   support

           0       0.61      0.48      0.54      2546
           1       0.56      0.69      0.62      2480

    accuracy                           0.58      5026
   macro avg       0.59      0.58      0.58      5026
weighted avg       0.59      0.58      0.58      5026

y1


Accuracy: 0.6287257451490298


              precision    recall  f1-score   support

           0       0.67      0.51      0.58      2487
           1       0.61      0.75      0.67      2512

    accuracy                           0.63      4999
   macro avg       0.64      0.63      0.62      4999
weighted avg       0.64      0.63      0.62      4999

y2


Accuracy: 0.5752450509321545


              precision    recall  f1-score   support

           0       0.58      0.51      0.54      2574
           1       0.57      0.64      0.60      2629

    accuracy                           0.58      5203
   macro avg       0

kernel=rbf c=1 gamma=1

Accuracy: 0.5237774030354132


Column: y0
              precision    recall  f1-score   support

           0       0.90      0.50      0.64      2503
           1       0.20      0.69      0.31       462

    accuracy                           0.53      2965
   macro avg       0.55      0.60      0.48      2965
weighted avg       0.79      0.53      0.59      2965

Column: y1
              precision    recall  f1-score   support

           0       0.90      0.52      0.66      2499
           1       0.21      0.68      0.32       466

    accuracy                           0.54      2965
   macro avg       0.55      0.60      0.49      2965
weighted avg       0.79      0.54      0.60      2965

Column: y2
              precision    recall  f1-score   support

           0       0.91      0.51      0.65      2609
           1       0.15      0.65      0.25       356

    accuracy                           0.53      2965
   macro avg       0.53      0.58      0.45      2965
weighted avg       0.82      0.53      0.61      2965

Column: y3
              precision    recall  f1-score   support

           0       0.86      0.45      0.59      2367
           1       0.24      0.70      0.36       598

    accuracy                           0.50      2965
   macro avg       0.55      0.57      0.47      2965
weighted avg       0.73      0.50      0.54      2965

Column: y4
              precision    recall  f1-score   support

           0       0.95      0.55      0.69      2711
           1       0.12      0.69      0.21       254

    accuracy                           0.56      2965
   macro avg       0.54      0.62      0.45      2965
weighted avg       0.88      0.56      0.65      2965

Column: y5
              precision    recall  f1-score   support

           0       0.80      0.38      0.52      2211
           1       0.29      0.73      0.41       754

    accuracy                           0.47      2965
   macro avg       0.55      0.55      0.46      2965
weighted avg       0.67      0.47      0.49      2965

Column: y6
              precision    recall  f1-score   support

           0       0.80      0.47      0.59      2216
           1       0.29      0.66      0.41       749

    accuracy                           0.52      2965
   macro avg       0.55      0.56      0.50      2965
weighted avg       0.67      0.52      0.54      2965

Column: y7
              precision    recall  f1-score   support

           0       0.92      0.55      0.69      2590
           1       0.17      0.65      0.27       375

    accuracy                           0.56      2965
   macro avg       0.54      0.60      0.48      2965
weighted avg       0.82      0.56      0.64      2965

Column: y8
              precision    recall  f1-score   support

           0       0.92      0.47      0.63      2604
           1       0.16      0.71      0.26       361

    accuracy                           0.50      2965
   macro avg       0.54      0.59      0.44      2965
weighted avg       0.83      0.50      0.58      2965

Column: y9
              precision    recall  f1-score   support

           0       0.86      0.51      0.64      2405
           1       0.23      0.64      0.34       560

    accuracy                           0.53      2965
   macro avg       0.55      0.57      0.49      2965
weighted avg       0.74      0.53      0.58      2965